File: 14-LongestCommonPrefix/main.go
package main

import "fmt"

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	if len(strs) == 1 {
		return strs[1]
	}

	prefix := strs[0]
	minLen := len(prefix)
	for _, str := range strs {
		if len(str) < minLen {
			minLen = len(str)
		}
	}

	for i := 0; i < minLen; i++ {
		char := prefix[i]
		for j := 1; j < len(strs); j++ {
			if strs[j][i] != char {
				return prefix[:i]
			}
		}
	}

	return prefix[:minLen]
}

func main() {
	fmt.Println(longestCommonPrefix([]string{"ab", "a", "abc"}))
	fmt.Println(longestCommonPrefix([]string{"flower", "flow", "flight"}))
	fmt.Println(longestCommonPrefix([]string{"dog", "racecar", "car"}))
}

File: 14-LongestCommonPrefix/main.py
from typing import List

class Solution:
  def longest_common_prefix(self, strs: List[str]) -> str: 
    if len(strs) == 0:
      return ""
    
    prefix = strs[0]
    if len(strs) == 1:
      return prefix 
    
    min_len = len(prefix)
    for str in strs:
      if len(str) < min_len:
        min_len = len(str)
    
    for i in range(min_len):
      char = prefix[i]
      for j in range(1, len(strs)):
        if char != strs[j][i]:
          return prefix[:i]
      
    return prefix[:min_len]

s = Solution()
print(s.longest_common_prefix(["flow", "flower", "flight"]))
print(s.longest_common_prefix(["a", "ab", "abc"]))
print(s.longest_common_prefix(["dog", "racecar", "car"]))